### PR TITLE
Made BaseTheme Progressive Web App (PWA)

### PR DESCRIPTION
**Pull request by BilelJegham**
Unsplash, Google Drive makes its websites PWA compatible. 
This pull request will make Roadiz BaseTheme Pwa compatible.

Add :
- [x] Add Manifest
- [x] Add Service Worker
- [x] Fix NSlink (Sitemap)

PWA audit Google Chrome

https://developers.google.com/web/progressive-web-apps/

### CODE REVIEW

**Comment by ambroisemaupate** [comment]
Thanks @BilelJegham that’s a great pull-request! 
I’ll review it as soon as possible.

I’ve changed it to our develop branch. We are using git flow as release workflow, so your code will be available on master in the next BaseTheme release.

File: Resources/views/sitemap/sitemap.xml.twig [hunk 1]
```diff
@@ -3,7 +3,9 @@
 
 {#  {% include '@BaseTheme/sitemap/sitemap-element.xml.twig' with {'element': home} only %}#}
 
-{% for page in pages if page.translation.Available %}
+
+
+{% for page in pages if page.translation.Available and page.nodeTypename != 'NSLink' %}
```
Comment (ambroisemaupate) [inline Resources/views/sitemap/sitemap.xml.twig]: `if` in `for` are now deprecated in Twig 2.x.

https://twig.symfony.com/doc/2.x/deprecated.html

Comment (BilelJegham) [reply Resources/views/sitemap/sitemap.xml.twig]: Thx, I learn something.

I fix this with `filter`

```twig

{% for page in pages|filter(p => p.translation.Available and page.nodeType.name != 'NSLink') %}
   
```

File: Resources/views/service-worker/sw.js.twig [hunk 2]
```diff
@@ -0,0 +1,46 @@
+importScripts('https://storage.googleapis.com/workbox-cdn/releases/4.3.1/workbox-sw.js');
+
+var CACHE_NAME = "cache-roadiz"
+const CACHE_ONE_WEEK = 7 * 24 * 60 * 60;
+const OFFLINE_URL = '/offline.html';
+const PAGES_TO_PRELOAD = [
+{% for page in pages|filter(p => p.translation.Available and page.nodeType.name != 'NSLink') %}
```
[the remark this answers is not in the document]
Comment (ambroisemaupate) [reply Resources/views/service-worker/sw.js.twig]: `{% for page in pages|filter(p => p.translation.Available) %}`

File: Resources/views/sitemap/sitemap.xml.twig [hunk 3]
```diff
@@ -3,7 +3,9 @@
 
 {#  {% include '@BaseTheme/sitemap/sitemap-element.xml.twig' with {'element': home} only %}#}
 
-{% for page in pages if page.translation.Available %}
+
+
+{% for page in pages|filter(p => p.translation.Available and page.nodeType.name != 'NSLink') %}
```
Comment (ambroisemaupate) [inline Resources/views/sitemap/sitemap.xml.twig]: `{% for page in pages|filter(p => p.translation.Available) %}`

File: Controllers/SitemapController.php [hunk 4]
```diff
@@ -40,7 +40,14 @@ public function sitemapAction(
             ->findBy([
                 'reachable' => true
```
Comment (ambroisemaupate) [inline Controllers/SitemapController.php]: ` 'name' => ['!=', 'Link'],`

File: Controllers/SitemapController.php [hunk 5]
```diff
@@ -40,7 +40,14 @@ public function sitemapAction(
...
             ]);
-
+        
+        $nodesLink = $this->get('em')
+             ->getRepository(NodeType::class)
+             ->findBy([
+                 'name' => 'Link'
+             ]); 
+        
+        $nodeTypes = array_diff($nodeTypes, $nodesLink);
```
Comment (ambroisemaupate) [inline Controllers/SitemapController.php]: Not necessary if you filter by name above

File: app/js/main.js [hunk 6]
```diff
@@ -122,3 +122,15 @@ const loadSvg = (r) => {
     nav.init()
     startingBlocks.boot()
 })()
+
+if ('serviceWorker' in navigator) {
```
Comment (ambroisemaupate) [inline app/js/main.js]: This condition does not work in Chrome 77 in http

File: app/js/main.js [hunk 7]
```diff
@@ -122,3 +122,15 @@ const loadSvg = (r) => {
...
+    window.addEventListener('load', function () {
+        navigator.serviceWorker.register('/sw.js').then(function (registration) {
+            // Registration was successful
+            console.log('ServiceWorker registration successful with scope: ', registration.scope)
+        }, function (err) {
+            // registration failed :(
+            console.log('ServiceWorker registration failed: ', err)
+        })
+    })
```
Comment (ambroisemaupate) [inline app/js/main.js]: You should add this : 

```js
 else if (!window.isSecureContext) {
    console.debug(
        'Navigator cannot init ServiceWorker, you should have a HTTPS secure context.',
        'Or force it unsafe chrome://flags/#unsafely-treat-insecure-origin-as-secure'
    )
}
```

File: Resources/views/service-worker/sw.js.twig (hunk 2 above)
[the remark this answers is not in the document]
Comment (ambroisemaupate) [reply Resources/views/service-worker/sw.js.twig]: I think you cannot use `and` in your filter condition.

File: app/js/main.js [hunk 8]
```diff
@@ -122,3 +122,15 @@ const loadSvg = (r) => {
     nav.init()
     startingBlocks.boot()
 })()
+
+if ('serviceWorker' in navigator) {
+    window.addEventListener('load', function () {
+        navigator.serviceWorker.register('/sw.js').then(function (registration) {
```
Comment (ambroisemaupate) [inline app/js/main.js]: Service worker path should not be hard coded.

Use a global var and generate this URL from twig.

Comment (ambroisemaupate) [reply app/js/main.js]:
```js
if (window.temp.swPath) {
    if ('serviceWorker' in navigator) {
        console.debug('Navigator support ServiceWorker')
        window.addEventListener('load', function () {
            navigator.serviceWorker.register(window.temp.swPath).then(function (registration) {
                // Registration was successful
                console.log('ServiceWorker registration successful with scope: ', registration.scope)
            }, function (err) {
                // registration failed :(
                console.log('ServiceWorker registration failed: ', err)
            })
        })
    } else if (!window.isSecureContext) {
        console.debug(
            'Navigator cannot init ServiceWorker, you should have a HTTPS secure context.',
            'Or force it unsafe chrome://flags/#unsafely-treat-insecure-origin-as-secure'
        )
    }
}
```

**Comment by BilelJegham** [comment]
> Hi @BilelJegham, could you allow me to commit on your pull request ?
> 
> I’ve fixed several issues from my last review.
> 
> image

image
Normally, @ambroisemaupate you are allowed to commit. 

Thanks for your review et comment.

**Comment by ambroisemaupate** [comment]
👏 Thank you @BilelJegham  and congrats for your first contribution on Roadiz’ BaseTheme.

**Pull request merged**